compiler/noirc_frontend/src/tests.rs

```diff
@@ -1911,25 +1917,62 @@ fn normal_generic_used_when_numeric_expected_in_where_clause() {
     }
     "#;
     let errors = get_program_errors(src);
-    assert_eq!(errors.len(), 1);
+    assert_eq!(errors.len(), 4);
     assert!(matches!(
         errors[0].0,
+        CompilationError::TypeError(TypeCheckError::TypeKindMismatch { .. }),
+    ));
+    assert!(matches!(
+        errors[1].0,
+        CompilationError::TypeError(TypeCheckError::TypeKindMismatch { .. }),
+    ));
+    assert!(matches!(
+        errors[2].0,
+        CompilationError::TypeError(TypeCheckError::TypeKindMismatch { .. }),
+    ));
+    // N
+    assert!(matches!(
+        errors[3].0,
         CompilationError::ResolverError(ResolverError::VariableNotDeclared { .. }),
     ));
 }
 
-// TODO(https://github.com/noir-lang/noir/issues/5156): Remove this test once we ban implicit numeric generics
 #[test]
-fn implicit_numeric_generics_elaborator() {
+fn implicit_numeric_generics_type_kind_mismatch() {
     let src = r#"
-    struct BoundedVec<T, MaxLen> {
+    fn foo<let N: u32>() -> u16 {
+        N as u16
+    }
+
+    global J: u16 = 10;
+
+    fn bar<let N: u16>() -> u16 {
+        foo::<J>()
+    }
+
+    global M: u16 = 3;
+    
+    fn main() {
+        let _ = bar::<M>();
+    }
+    "#;
+    let errors = get_program_errors(src);
+    assert_eq!(errors.len(), 1);
+    assert!(matches!(
+        errors[0].0,
+        CompilationError::TypeError(TypeCheckError::TypeKindMismatch { .. }),
+    ));
+}
+
+#[test]
+fn implicit_numeric_generics_value_kind_mismatch() {
+    let src = r#"
+    struct BoundedVec<T, let MaxLen: u32> {
         storage: [T; MaxLen],
         len: u64,
     }
-    
-    impl<T, MaxLen> BoundedVec<T, MaxLen> {
 
-        // Test that we have an implicit numeric generic for "Len" as well as "MaxLen"
+    impl<T, let MaxLen: u32> BoundedVec<T, MaxLen> {
         pub fn extend_from_bounded_vec<Len>(&mut self, _vec: BoundedVec<T, Len>) { 
```

Comment:
Let's be more explicit about this type kind mismatch by making it a `u64` of `Field`. I missed this in my last review. I think this may be why we're also getting a `VariableNotDeclared` error.